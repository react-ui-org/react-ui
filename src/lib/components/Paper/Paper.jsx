import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import styles from './Paper.scss';

export const Paper = ({
  children,
  id,
  muted,
  raised,
}) => (
  <div
    className={classNames(
      styles.root,
      muted && styles.rootMuted,
      raised && styles.rootRaised,
    )}
    id={id}
  >
    {children}
  </div>
);

Paper.defaultProps = {
  children: null,
  id: undefined,
  muted: false,
  raised: false,
};

Paper.propTypes = {
  /**
   * Content to be placed onto Paper.
   */
  children: PropTypes.node,
  /**
   * ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Visually suppress Paper.
   */
  muted: PropTypes.bool,
  /**
   * Add shadow to pull the Paper above surface.
   */
  raised: PropTypes.bool,
};

export const PaperWithContext = withProviderContext(Paper, 'Paper');

export default PaperWithContext;
