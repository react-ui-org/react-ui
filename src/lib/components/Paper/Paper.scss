@use "theme";

.root {
    padding: theme.$padding;
    border: theme.$border-width solid theme.$border-color;
    border-radius: theme.$border-radius;
    background-color: theme.$background-color;
}

.rootMuted {
    background-color: theme.$muted-background-color;
    opacity: theme.$muted-opacity;
}

.rootRaised {
    box-shadow: theme.$raised-box-shadow;
}
