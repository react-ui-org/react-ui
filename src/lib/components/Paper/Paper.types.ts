export interface PaperProps {
  /**
   * Rest of the props.
   */
  [key: string]: unknown;
  /**
   * Content to be placed onto Paper.
   */
  children: React.ReactNode;
  /**
   * Visually suppress Paper.
   */
  muted?: false;
  /**
   * Add shadow to pull the Paper above surface.
   */
  raised?: false;
}
