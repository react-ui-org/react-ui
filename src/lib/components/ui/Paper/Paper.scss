@use 'theme';

.root {
  display: flex;
  flex-direction: column;
  width: 100%;
  padding: theme.$padding;
  border: theme.$border-width solid theme.$border-color;
  border-radius: theme.$border-radius;
  background-color: theme.$background-color;
}

.rootRaised {
  box-shadow: theme.$raised-box-shadow;
}
