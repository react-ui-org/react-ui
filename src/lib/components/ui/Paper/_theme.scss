$padding: var(--rui-Paper__padding);
$border-width: var(--rui-Paper__border-width);
$border-color: var(--rui-Paper__border-color);
$border-radius: var(--rui-Paper__border-radius);
$background-color: var(--rui-Paper__background-color);
$raised-box-shadow: var(--rui-Paper--raised__box-shadow);
