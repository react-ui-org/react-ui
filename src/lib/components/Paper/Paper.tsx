import React from 'react';
import { withGlobalProps } from '../../provider';
import { classNames } from '../../utils/classNames';
import { transferProps } from '../_helpers/transferProps';
import styles from './Paper.scss';
import { PaperProps } from './Paper.types';

export const Paper: React.FunctionComponent<PaperProps> = ({
  children,
  muted = false,
  raised = false,
  ...restProps
}) => (
  <div
    {...transferProps(restProps)}
    className={classNames(
      styles.root,
      muted && styles.isRootMuted,
      raised && styles.isRootRaised,
    )}
  >
    {children}
  </div>
);

export const PaperWithGlobalProps = withGlobalProps(Paper, 'Paper');

export default PaperWithGlobalProps;
