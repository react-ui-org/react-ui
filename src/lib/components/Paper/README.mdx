---
name: Paper
menu: 'Miscellaneous'
route: /components/paper
---

# Paper

import {
  Playground,
  Props,
} from 'docz'
import { Paper } from './Paper'

Paper is a basic surface to hold content.

## Basic Usage

To implement the Paper component, you need to import it first:

```js
import { Paper } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <Paper>
    Hello!
  </Paper>
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Paper is **designed for non-white background.** You may want to either use its
  [raised variant](#raised-paper) or [customize](/customize/theming/overview)
  the default appearance to make it stand out on white background.

- **Paper, or Card?** Paper is a basic surface to put content on. However,
  there is also the [Card](/components/card) component. While Paper is
  usually used to hold larger content areas like lists, grids, or forms, Card is
  designed for displaying items. Card also supports more visual options.

## Raised Paper

Add optional shadow to lift the paper above background.

<Playground>
  <Paper raised>
    Hello! I&apos;m paper and I&apos;m raised.
  </Paper>
</Playground>

## Muted Paper

Dim background and add transparency to visually suppress the Paper.

<Playground>
  <Paper muted>
    Sssh! I&apos;m paper and I&apos;m muted.
  </Paper>
</Playground>

## API

<Props table of={Paper} />

## Theming

| Custom Property                                      | Description                                                  |
|------------------------------------------------------|--------------------------------------------------------------|
| `--rui-Paper__padding`                               | Padding of Paper                                             |
| `--rui-Paper__border-width`                          | Border width                                                 |
| `--rui-Paper__border-color`                          | Border color                                                 |
| `--rui-Paper__border-radius`                         | Corner radius                                                |
| `--rui-Paper__background-color`                      | Paper background color                                       |
| `--rui-Paper--muted__background-color`               | Background color of muted paper                              |
| `--rui-Paper--muted__opacity`                        | Opacity of muted paper                                       |
| `--rui-Paper--raised__box-shadow`                    | Box shadow of raised paper                                   |
