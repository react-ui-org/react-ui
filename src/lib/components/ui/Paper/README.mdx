---
name: Paper
menu: 'UI'
route: /components/ui/paper
---

# Paper

import {
  Playground,
  Props,
} from 'docz'
import { Paper } from './Paper'

Paper is a basic surface to hold content.

## Basic Usage

To implement the Paper component, you need to import it first:

```js
import { Paper } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <Paper>
    Hello!
  </Paper>
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Paper is **designed for non-white background.** You may want to either use its
  [raised variant](#raised-paper) or [customize](/customize/theming/overview)
  the default appearance to make it stand out on white background.

- **Paper, or Card?** Paper is a basic surface to put content on. However,
  there is also the [Card](/components/ui/card) component. While Paper is
  usually used to hold larger content areas like lists, grids, or forms, Card is
  designed for displaying items. Card also supports more visual options.

## Raised Paper

Add optional shadow to lift the paper above background.

<Playground>
  <Paper raised>
    Hello! I&apos;m paper and I&apos;m raised.
  </Paper>
</Playground>

## API

<Props table of={Paper} />

## Theming

| Custom Property                                      | Description                                                  |
|------------------------------------------------------|--------------------------------------------------------------|
| `--rui-Paper__padding`                               | Padding of Paper                                             |
| `--rui-Paper__border-width`                          | Border width                                                 |
| `--rui-Paper__border-color`                          | Border color                                                 |
| `--rui-Paper__border-radius`                         | Corner radius                                                |
| `--rui-Paper__background-color`                      | Paper background color                                       |
| `--rui-Paper--raised__box-shadow`                    | Box shadow of raised paper                                   |
