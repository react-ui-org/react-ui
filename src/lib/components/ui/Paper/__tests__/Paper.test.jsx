import React from 'react';
import { shallow } from 'enzyme';
import { Paper } from '../Paper';

describe('rendering', () => {
  it('renders correctly', () => {
    const tree = shallow(<Paper><p>Paper content</p></Paper>);

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(
      <Paper id="custom-id" raised>
        <p>Paper content</p>
      </Paper>,
    );

    expect(tree).toMatchSnapshot();
  });
});
