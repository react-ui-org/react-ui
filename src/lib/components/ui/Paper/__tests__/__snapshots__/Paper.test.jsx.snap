// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<div
  className="root "
>
  <p>
    Paper content
  </p>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root rootRaised"
  id="custom-id"
>
  <p>
    Paper content
  </p>
</div>
`;
